chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,949 @@
+        );
+        store_update.set_flat_storage_status(
+            right_child_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CreatingChild),
+        );
+        store_update.commit()?;
+
+        self.schedule_split_shard(parent_shard, &status, scheduler, controller);
+        Ok(())
+    }
+
+    /// Returns an error if a resharding event is in progress.
+    fn check_no_resharding_in_progress(&self) -> Result<(), StorageError> {
+        // Do not allow multiple resharding events in parallel.
+        if self.resharding_event().is_some() {
+            error!(target: "resharding", "trying to start a new flat storage resharding event while one is already in progress!");
+            Err(StorageError::FlatStorageReshardingAlreadyInProgress)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn set_resharding_event(&self, event: FlatStorageReshardingEvent) {
+        *self.inner.resharding_event.lock().unwrap() = Some(event);
+    }
+
+    /// Returns the current in-progress resharding event, if any.
+    pub fn resharding_event(&self) -> Option<FlatStorageReshardingEvent> {
+        self.inner.resharding_event.lock().unwrap().clone()
+    }
+
+    /// Schedules a task to split a shard.
+    fn schedule_split_shard(
+        &self,
+        parent_shard: ShardUId,
+        status: &SplittingParentStatus,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) {
+        let event = FlatStorageReshardingEvent::Split(parent_shard, status.clone());
+        self.set_resharding_event(event);
+        info!(target: "resharding", ?parent_shard, ?status,"scheduling flat storage shard split");
+
+        let resharder = self.inner.clone();
+        let task = Box::new(move || split_shard_task(resharder, controller));
+        scheduler.schedule(task);
+    }
+
+    /// Cleans up children shards flat storage's content (status is excluded).
+    fn clean_children_shards(&self, status: &SplittingParentStatus) -> Result<(), Error> {
+        let SplittingParentStatus { left_child_shard, right_child_shard, .. } = status;
+        debug!(target: "resharding", ?left_child_shard, ?right_child_shard, "cleaning up children shards flat storage's content");
+        let mut store_update = self.inner.runtime.store().flat_store().store_update();
+        for child in [left_child_shard, right_child_shard] {
+            store_update.remove_all_deltas(*child);
+            store_update.remove_all(*child);
+        }
+        store_update.commit()?;
+        Ok(())
+    }
+}
+
+/// Struct used to destructure a new shard layout definition into the resulting resharding event.
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+enum ReshardingEventType {
+    /// Split of a shard.
+    Split(ReshardingSplitParams),
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+struct ReshardingSplitParams {
```

Comment:
`ReshardingSplitParams` is the output of processing the shard layout and the latter has already the boundary. Well, conceptually it would be nice to have the boundary in `ReshardingSplitParams` (I started with that), but I found the boundary to be pretty useless on its own: one needs the shard layout and the account-id to get the destination shard.

Reply:
Looking at the split method, it takes the params and the shard layout. Would it make sense to shove the shard layout inside of the params? You can then get rid of left/ right child because they can be obtained from the shard layout.